File: backend/app/domain/events/order_event_handlers.py
import json
import logging

from sqlalchemy import update

from app.core.notifications.notification_service import (
    Notification,
    NotificationChannel,
    NotificationService,
)
from app.domain.events.event_bus import get_event_bus
from app.domain.events.order_events import (
    OrderCancelledEvent,
    OrderCreatedEvent,
    OrderDeliveredEvent,
    OrderShippedEvent,
    OrderStatusChangedEvent,
    PaymentProcessedEvent,
)
from app.models.product import Product
from app.services.order_service import OrderService

logger = logging.getLogger(__name__)

"""
Developer Note: Event Flow
-------------------------
All order lifecycle changes (creation, payment, status change, cancellation, refund) emit domain events via the event bus.
Handlers are registered for each event type to perform side effects (notifications, analytics, inventory deduction, etc.).
Never update order status or perform side effects directly in service logic—always use events and handlers for extensibility and auditability.
"""


def analytics_log_event(event_type, **kwargs):
    """Log structured analytics event as JSON."""
    log_data = {"event_type": event_type, **kwargs}
    logger.info("[Analytics] %s", json.dumps(log_data))


async def handle_order_created(event: OrderCreatedEvent):
    # 1. Send notification (email/SMS/WhatsApp)
    notification = Notification(
        id=event.event_id,
        tenant_id=event.tenant_id,
        user_id=event.order.customer.id or "",
        title="Order Confirmation",
        message=f"Your order #{event.order.order_number} has been created!",
        priority="medium",
        channels=[NotificationChannel.EMAIL, NotificationChannel.SMS],
        metadata={"order_id": event.order_id},
    )
    service = NotificationService()
    try:
        await service.send_notification(notification)
    except Exception as e:
        logger.error(
            f"Notification sending failed for order {event.order_id}: {str(e)}"
        )

    # 2. Log analytics (placeholder: log to logger)
    analytics_log_event(
        "OrderCreated",
        order_id=event.order_id,
        value=event.order.total_amount.amount,
        currency=event.order.total_amount.currency,
    )

    # 3. Trigger fulfillment workflow (placeholder)
    logger.info(
        f"[Fulfillment] Trigger fulfillment for order {event.order_id}")


async def handle_order_status_changed(event: OrderStatusChangedEvent):
    # 1. Send notification (email/SMS/WhatsApp)
    notification = Notification(
        id=event.event_id,
        tenant_id=event.tenant_id,
        user_id=event.changed_by or "",
        title="Order Status Updated",
        message=f"Your order #{event.order_number} status changed: {event.previous_status} → {event.new_status}",
        priority="medium",
        channels=[NotificationChannel.EMAIL, NotificationChannel.SMS],
        metadata={"order_id": event.order_id, "new_status": event.new_status},
    )
    service = NotificationService()
    try:
        await service.send_notification(notification)
    except Exception as e:
        logger.error(
            f"Notification sending failed for order {event.order_id}: {str(e)}"
        )

    # 2. Log analytics
    analytics_log_event(
        "OrderStatusChanged",
        order_id=event.order_id,
        previous_status=event.previous_status,
        new_status=event.new_status,
    )

    # 3. Fulfillment/side effects
    logger.info(
        f"[Fulfillment] Handle status change for order {event.order_id} to {event.new_status}"
    )


async def handle_order_shipped(event: OrderShippedEvent):
    # 1. Send notification
    notification = Notification(
        id=event.event_id,
        tenant_id=event.tenant_id,
        user_id="",
        title="Order Shipped",
        message=f"Your order #{event.order_number} has shipped! Tracking: {event.tracking_number}",
        priority="medium",
        channels=[NotificationChannel.EMAIL, NotificationChannel.SMS],
        metadata={"order_id": event.order_id,
                  "tracking_number": event.tracking_number},
    )
    service = NotificationService()
    try:
        await service.send_notification(notification)
    except Exception as e:
        logger.error(
            f"Notification sending failed for order {event.order_id}: {str(e)}"
        )

    # 2. Log analytics
    analytics_log_event(
        "OrderShipped", order_id=event.order_id, tracking_number=event.tracking_number
    )

    # 3. Fulfillment/side effects
    logger.info(
        f"[Fulfillment] Notify warehouse/shipping provider for order {event.order_id}"
    )


async def handle_order_delivered(event: OrderDeliveredEvent):
    # 1. Send notification
    notification = Notification(
        id=event.event_id,
        tenant_id=event.tenant_id,
        user_id="",
        title="Order Delivered",
        message=f"Your order #{event.order_number} was delivered on {event.delivery_date}",
        priority="medium",
        channels=[NotificationChannel.EMAIL, NotificationChannel.SMS],
        metadata={
            "order_id": event.order_id,
            "delivery_date": str(event.delivery_date),
        },
    )
    service = NotificationService()
    try:
        await service.send_notification(notification)
    except Exception as e:
        logger.error(
            f"Notification sending failed for order {event.order_id}: {str(e)}"
        )

    # 2. Log analytics
    analytics_log_event(
        "OrderDelivered", order_id=event.order_id, delivery_date=event.delivery_date
    )

    # 3. Fulfillment/side effects
    logger.info(
        f"[Fulfillment] Complete fulfillment for order {event.order_id}")


async def handle_order_cancelled(event: OrderCancelledEvent):
    # 1. Send notification
    notification = Notification(
        id=event.event_id,
        tenant_id=event.tenant_id,
        user_id=event.cancelled_by or "",
        title="Order Cancelled",
        message=f"Your order #{event.order_number} was cancelled. Reason: {event.cancellation_reason}",
        priority="medium",
        channels=[NotificationChannel.EMAIL, NotificationChannel.SMS],
        metadata={"order_id": event.order_id,
                  "reason": event.cancellation_reason},
    )
    service = NotificationService()
    try:
        await service.send_notification(notification)
    except Exception as e:
        logger.error(
            f"Notification sending failed for order {event.order_id}: {str(e)}"
        )

    # 2. Log analytics
    analytics_log_event(
        "OrderCancelled", order_id=event.order_id, reason=event.cancellation_reason
    )

    # 3. Fulfillment/side effects
    logger.info(
        f"[Fulfillment] Cancel fulfillment/refund for order {event.order_id}")


async def handle_payment_processed(event: PaymentProcessedEvent):
    # 1. Send notification
    notification = Notification(
        id=event.event_id,
        tenant_id=event.tenant_id,
        user_id="",
        title="Payment Processed",
        message=f"Payment for order #{event.order_number} was processed. Amount: {event.amount}",
        priority="medium",
        channels=[NotificationChannel.EMAIL, NotificationChannel.SMS],
        metadata={"order_id": event.order_id,
                  "transaction_id": event.transaction_id},
    )
    service = NotificationService()
    try:
        await service.send_notification(notification)
    except Exception as e:
        logger.error(
            f"Notification sending failed for order {event.order_id}: {str(e)}"
        )

    # 2. Log analytics
    analytics_log_event(
        "PaymentProcessed",
        order_id=event.order_id,
        transaction_id=event.transaction_id,
        amount=event.amount,
    )

    # 3. Fulfillment/side effects
    logger.info(
        f"[Fulfillment] Release order for fulfillment after payment {event.order_id}"
    )


async def handle_inventory_deduction(event):
    """Deduct inventory for products in the order when an order is created or paid."""
    db = (
        event.order._sa_instance_state.session
        if hasattr(event.order, "_sa_instance_state")
        else None
    )
    if not db:
        logging.error("No DB session found for inventory deduction.")
        return
    try:
        for item in getattr(event.order, "items", []):
            product_id = getattr(item, "product_id", None)
            quantity = getattr(item, "quantity", 1)
            if product_id:
                await db.execute(
                    update(Product)
                    .where(Product.id == product_id)
                    .values(inventory_quantity=Product.inventory_quantity - quantity)
                )
        await db.commit()
        logging.info(f"Inventory deducted for order {event.order.id}")
    except Exception as e:
        await db.rollback()
        logging.error(
            f"Inventory deduction failed for order {event.order.id}: {str(e)}"
        )


async def handle_fulfillment(event):
    """Simulate fulfillment workflow integration for shipping and delivery events."""
    if hasattr(event, "order_id"):
        if getattr(event, "event_type", None) == "ORDER_SHIPPED":
            logger.info(
                f"[Fulfillment] Order {event.order_id} shipped. Notifying warehouse/shipping provider."
            )
            # Simulate integration with fulfillment provider here
        elif getattr(event, "event_type", None) == "ORDER_DELIVERED":
            logger.info(
                f"[Fulfillment] Order {event.order_id} delivered. Completing fulfillment."
            )
            # Simulate post-delivery actions here


# Register the handlers
get_event_bus().subscribe("ORDER_CREATED", handle_order_created)
get_event_bus().subscribe("ORDER_STATUS_CHANGED", handle_order_status_changed)
get_event_bus().subscribe("ORDER_SHIPPED", handle_order_shipped)
get_event_bus().subscribe("ORDER_DELIVERED", handle_order_delivered)
get_event_bus().subscribe("ORDER_CANCELLED", handle_order_cancelled)
get_event_bus().subscribe("PAYMENT_PROCESSED", handle_payment_processed)
get_event_bus().subscribe("ORDER_CREATED", handle_inventory_deduction)
get_event_bus().subscribe("PAYMENT_PROCESSED", handle_inventory_deduction)
get_event_bus().subscribe("ORDER_SHIPPED", handle_fulfillment)
get_event_bus().subscribe("ORDER_DELIVERED", handle_fulfillment)

"""
Fulfillment Orchestration:
- Fulfillment steps (shipping, delivery) are triggered by events (ORDER_SHIPPED, ORDER_DELIVERED).
- The handle_fulfillment handler simulates integration with a fulfillment provider (e.g., warehouse, shipping API).
- To extend, implement real API calls or queue jobs in this handler.
"""

File: frontend/src/modules/order/events/OrderEventHandlers.ts
import type {
  OrderCreatedEvent,
  OrderStatusChangedEvent,
  PaymentProcessedEvent,
  OrderShippedEvent,
  OrderDeliveredEvent,
  OrderCancelledEvent,
} from '@/modules/order/models/events/orderEvents';

import { getEventBus } from './EventBus';

/**
 * Initialize order event handlers
 * Registers handlers for different order events to execute side-effects
 */
export function initializeOrderEventHandlers(): () => void {
  const eventBus = getEventBus();

  // Create individual subscriptions
  const unsubscribeOrderCreated = eventBus.subscribe<OrderCreatedEvent>(
    'ORDER_CREATED',
    handleOrderCreated,
  );

  const unsubscribeStatusChanged = eventBus.subscribe<OrderStatusChangedEvent>(
    'ORDER_STATUS_CHANGED',
    handleOrderStatusChanged,
  );

  const unsubscribePaymentProcessed = eventBus.subscribe<PaymentProcessedEvent>(
    'PAYMENT_PROCESSED',
    handlePaymentProcessed,
  );

  const unsubscribeOrderShipped = eventBus.subscribe<OrderShippedEvent>(
    'ORDER_SHIPPED',
    handleOrderShipped,
  );

  const unsubscribeOrderDelivered = eventBus.subscribe<OrderDeliveredEvent>(
    'ORDER_DELIVERED',
    handleOrderDelivered,
  );

  const unsubscribeOrderCancelled = eventBus.subscribe<OrderCancelledEvent>(
    'ORDER_CANCELLED',
    handleOrderCancelled,
  );

  // Return combined unsubscribe function
  return () => {
    unsubscribeOrderCreated();
    unsubscribeStatusChanged();
    unsubscribePaymentProcessed();
    unsubscribeOrderShipped();
    unsubscribeOrderDelivered();
    unsubscribeOrderCancelled();
  };
}

/**
 * Handle order created events
 * - Update UI
 * - Show notifications
 * - Update analytics
 */
async function handleOrderCreated(event: OrderCreatedEvent): Promise<void> {
  console.log('[OrderEvent] Order created:', event.order_number);

  try {
    // Show notification
    if (typeof window !== 'undefined' && 'Notification' in window) {
      showNotification('Order Created', `Your order #${event.order_number} has been created`);
    }

    // Track analytics
    trackOrderEvent('order_created', {
      order_id: event.order_id,
      order_number: event.order_number,
      value: event.order.total_amount.amount,
      currency: event.order.total_amount.currency,
      items: event.order.items.length,
    });

    // Additional side effects...
  } catch (error) {
    console.error('Error handling order created event:', error);
  }
}

/**
 * Handle order status changed events
 * - Update UI
 * - Show notifications
 */
async function handleOrderStatusChanged(event: OrderStatusChangedEvent): Promise<void> {
  console.log(
    `[OrderEvent] Order ${event.order_number} status changed: ${event.previous_status} -> ${event.new_status}`,
  );

  try {
    // Show notification
    if (typeof window !== 'undefined' && 'Notification' in window) {
      showNotification(
        'Order Status Updated',
        `Your order #${event.order_number} is now ${event.new_status}`,
      );
    }

    // Track status change
    trackOrderEvent('order_status_changed', {
      order_id: event.order_id,
      order_number: event.order_number,
      previous_status: event.previous_status,
      new_status: event.new_status,
    });

    // Additional side effects based on the specific status...
  } catch (error) {
    console.error('Error handling order status changed event:', error);
  }
}

/**
 * Handle payment processed events
 */
async function handlePaymentProcessed(event: PaymentProcessedEvent): Promise<void> {
  console.log(
    `[OrderEvent] Payment processed for order ${event.order_number}: ${event.payment_status}`,
  );

  try {
    // Show notification
    if (typeof window !== 'undefined' && 'Notification' in window) {
      const message =
        event.payment_status === 'COMPLETED'
          ? `Payment of ${event.amount.amount} ${event.amount.currency} was successful`
          : `Payment for order #${event.order_number} failed`;

      showNotification('Payment Update', message);
    }

    // Track payment
    trackOrderEvent('payment_processed', {
      order_id: event.order_id,
      order_number: event.order_number,
      status: event.payment_status,
      amount: event.amount.amount,
      currency: event.amount.currency,
      payment_method: event.payment_method,
    });
  } catch (error) {
    console.error('Error handling payment processed event:', error);
  }
}

/**
 * Handle order shipped events
 */
async function handleOrderShipped(event: OrderShippedEvent): Promise<void> {
  console.log(`[OrderEvent] Order ${event.order_number} shipped`);

  try {
    // Show notification with tracking info
    if (typeof window !== 'undefined' && 'Notification' in window) {
      const message = event.tracking_number
        ? `Your order #${event.order_number} has shipped. Tracking: ${event.tracking_number}`
        : `Your order #${event.order_number} has shipped`;

      showNotification('Order Shipped', message);
    }

    // Track shipment
    trackOrderEvent('order_shipped', {
      order_id: event.order_id,
      order_number: event.order_number,
      tracking_number: event.tracking_number,
      shipping_provider: event.shipping_provider,
    });
  } catch (error) {
    console.error('Error handling order shipped event:', error);
  }
}

/**
 * Handle order delivered events
 */
async function handleOrderDelivered(event: OrderDeliveredEvent): Promise<void> {
  console.log(`[OrderEvent] Order ${event.order_number} delivered`);

  try {
    // Show notification
    if (typeof window !== 'undefined' && 'Notification' in window) {
      showNotification('Order Delivered', `Your order #${event.order_number} has been delivered!`);
    }

    // Track delivery
    trackOrderEvent('order_delivered', {
      order_id: event.order_id,
      order_number: event.order_number,
      delivery_date: event.delivery_date,
    });

    // Prompt for review after delivery
    setTimeout(() => {
      // Could show review prompt
    }, 3000);
  } catch (error) {
    console.error('Error handling order delivered event:', error);
  }
}

/**
 * Handle order cancelled events
 */
async function handleOrderCancelled(event: OrderCancelledEvent): Promise<void> {
  console.log(`[OrderEvent] Order ${event.order_number} cancelled`);

  try {
    // Show notification
    if (typeof window !== 'undefined' && 'Notification' in window) {
      const reason = event.cancellation_reason ? ` Reason: ${event.cancellation_reason}` : '';

      showNotification(
        'Order Cancelled',
        `Your order #${event.order_number} has been cancelled.${reason}`,
      );
    }

    // Track cancellation
    trackOrderEvent('order_cancelled', {
      order_id: event.order_id,
      order_number: event.order_number,
      reason: event.cancellation_reason,
      refund_initiated: event.refund_initiated,
    });
  } catch (error) {
    console.error('Error handling order cancelled event:', error);
  }
}

/**
 * Show a browser notification
 * With fallback to console for browsers without permission
 */
function showNotification(title: string, message: string): void {
  // Only run in browser environment
  if (typeof window === 'undefined') return;

  // Check if browser supports notifications
  if (!('Notification' in window)) {
    console.log(`${title}: ${message}`);
    return;
  }

  // Show notification if permission granted
  if (Notification.permission === 'granted') {
    new Notification(title, { body: message });
  }
  // Request permission if not determined
  else if (Notification.permission !== 'denied') {
    Notification.requestPermission().then((permission) => {
      if (permission === 'granted') {
        new Notification(title, { body: message });
      }
    });
  }
}

/**
 * Track order events for analytics
 * Implements offline resilience for African markets
 */
function trackOrderEvent(eventName: string, properties: Record<string, unknown>): void {
  try {
    // If analytics is available, track the event
    if (typeof window !== 'undefined' && window.analytics) {
      window.analytics.track(eventName, properties);
      return;
    }

    // Offline fallback - store events to send later
    if (typeof window !== 'undefined') {
      const events = JSON.parse(localStorage.getItem('offline_analytics') || '[]');
      events.push({
        event: eventName,
        properties,
        timestamp: new Date().toISOString(),
      });
      localStorage.setItem('offline_analytics', JSON.stringify(events));

      // Process offline events when online
      if (!window.__analyticsOnlineListener) {
        window.__analyticsOnlineListener = true;
        window.addEventListener('online', processOfflineAnalytics);
      }
    }
  } catch (error) {
    console.error('Error tracking order event:', error);
  }
}

/**
 * Process offline analytics events when connection is restored
 */
function processOfflineAnalytics(): void {
  if (typeof window === 'undefined' || !window.analytics) return;

  try {
    const events = JSON.parse(localStorage.getItem('offline_analytics') || '[]');
    if (events.length === 0) return;

    // Process each event
    events.forEach((event: { event: string; properties: Record<string, unknown>; timestamp: string }) => {
      window.analytics?.track(event.event, {
        ...event.properties,
        offline_tracked_at: event.timestamp,
      });
    });

    // Clear the offline events
    localStorage.setItem('offline_analytics', '[]');
  } catch (error) {
    console.error('Error processing offline analytics:', error);
  }
}

// Add global types
declare global {
  interface Window {
    analytics?: {
      track: (event: string, properties?: Record<string, unknown>) => void;
    };
    __analyticsOnlineListener?: boolean;
  }
}
